### PR TITLE
[diffusion] add tiered MiniMax-H3 AdaLN cache

### PR DESCRIPTION
**Pull request by Dayuxiaoshui**
Related issue: #35495

## Motivation

MiniMax-H3 can avoid keeping the 24.2 GiB `adaln_proj` checkpoint weights resident by rebuilding AdaLN outputs online. The previous online path kept a fixed GPU slab of historical timestep plans. That slab reserved memory for plans that the current request did not use and discarded all resident plans when its capacity was exceeded. Returning to a recently used schedule therefore required another full checkpoint scan.

This change adds a tiered cache manager that keeps only the current request's working set on GPU, stores complete schedules in a bounded pinned-CPU LRU, and rebuilds schedules from checkpoint weights only on CPU misses. Exact FP32 timestep-plan matching is preserved; approximate timestep merging and scheduler changes are out of scope.

## Modifications

- Replaced the historical GPU plan set with a request-local working set backed by stable GPU tensors. The tensor pointers remain stable for CUDA graph compatibility, while only the active request's plans are published through `plan_lengths`.
- Added exact schedule keys containing every timestep plan used by the request. A complete schedule is treated as one cache entry so partial hits cannot expose an inconsistent request.
- Added a byte-bounded pinned-CPU LRU cache for complete projected AdaLN schedules. CPU hits restore the timestep metadata and projected parameters to the GPU workspace without rereading checkpoint weights.
- Added transactional publication for rebuilds and host restores. A failed checkpoint read, projection, allocation, or shape validation leaves the previous request invisible and allows a later request to retry safely.
- Preserved TP sharding semantics by reading only the local AdaLN column shard and using the same all-gather behavior as the existing column-parallel path.
- Added GPU-hit, host-hit, rebuild, transfer, admission, eviction, and miss counters to the request perf dump through custom metrics.
- Added `--minimax-h3-adaln-max-plans` and `--minimax-h3-adaln-host-cache-gb`. The existing default remains `64` plans and width `4` for backwards compatibility; deployments can select a request-sized workspace explicitly, for example `7` plans and width `2` for T2VA or width `3` for FL2VA.
- Added validation for invalid plan capacity, plan width, and non-finite or negative host-cache sizes.
- Online AdaLN now automatically disables breakable CUDA graph capture at startup with an actionable warning because dynamic online rebuild and the current capture path are not replay-safe. Resident-weight and fixed-sidecar paths retain their existing graph behavior.
- Added unit coverage for GPU hits, host hits, byte-based eviction, transactional retry, request reset, pointer stability, shape/width rejection, perf metric serialization, and online-AdaLN graph compatibility.

## Accuracy Tests

- MiniMax-H3 AdaLN unit tests: `10 passed`.
- Server-argument and graph-compatibility tests: `13 passed`, including both online-AdaLN graph downgrade and sidecar graph-preservation cases.
- The audited AdaLN comparison covered `470,367,744` output elements at TP1 and TP2 with zero mismatches.
- A real TP2 MiniMax-H3 FL2VA run completed the full submit, denoise, video/audio decode, MP4 creation, and download path for an `A -> B -> A` schedule sequence. The first and third requests used the same prompt, seed, resolution, duration, and schedule; their MP4 SHA256 values were identical:
  - `91186630bf7862d3892877bd1bacaa4464b3937b5a1f45ee8190be58995104af`
- The third request was served from the pinned host tier and did not rebuild AdaLN outputs.

## Speed Tests and Profiling

Environment: `2 x H200`, MiniMax-H3 FL2VA, TP2, Ulysses 1, 1344x768, 4 seconds, 8 inference steps.

| Request | AdaLN preparation | Tier | E2E wall time |
| --- | ---: | --- | ---: |
| Schedule A, first request | `2753.104 ms` | rebuild | `65.166 s` |
| Schedule B | `2390.848 ms` | rebuild | `50.135 s` |
| Schedule A, return | `7.392 ms` | pinned host hit | `47.115 s` |

The host-hit request transferred `135,776,312` bytes to GPU memory with one H2D transfer, recorded `host_hits=1`, and kept `rebuilds=2` after the three-request sequence. The configured `7`-plan, width-`3` FL2VA workspace used approximately `0.19 GiB` per card for AdaLN buffers.

The cold-miss profile shows that checkpoint movement dominates rebuild cost: approximately `24.3 GiB` of AdaLN checkpoint data is transferred, while the main projection GEMMs account for only a small fraction of the capture. The pinned host transfer avoids that checkpoint scan on repeated schedules.

Validation commands completed:

- `python -m black --check` on all modified Python files.
- `python -m py_compile` on all modified Python modules and tests.
- `git diff --check`.

The pre-commit hook could not download its external environment because the execution environment had no GitHub DNS/network access; this is an environment limitation rather than a formatting failure.

## Checklist

- [x] Code formatting and whitespace checks passed with Black and `git diff --check`.
- [x] Added focused unit tests for cache tiers, transactional failure behavior, metrics, and graph compatibility.
- [x] Added this implementation and benchmark documentation.
- [x] Provided TP2 end-to-end accuracy and speed measurements.
- [x] Reviewed the implementation against SGLang code style and preserved existing sidecar/resident paths.

## Review and Merge Process

1. Ping Merge Oncalls after the PR has received the required CODEOWNERS reviews.
2. Trigger CI with `/tag-and-rerun-ci`, `/tag-run-ci-label`, or `/rerun-failed-ci` when appropriate.
3. Confirm the MiniMax-H3 unit tests and relevant multimodal generation CI jobs are green.
4. After approvals and green CI, ask a Merge Oncall or maintainer with write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #32327627623](https://github.com/sgl-project/sglang/actions/runs/32327627623)
Latest PR Test (Extra): :x: [Run #32327627456](https://github.com/sgl-project/sglang/actions/runs/32327627456)

### CODE REVIEW

**Comment by Dayuxiaoshui** [comment]
cc @mickqian @BBuf

**Review by triple-mu — COMMENTED** [review]
This PR changes the online AdaLN GPU cache from "accumulate plans across requests" to "keep only the current request's schedule", and adds a pinned host LRU behind it. The direction is right, but three things would be worth resolving before merge: TP>1 can mismatch collectives, the default configuration is a regression against today, and the stated reason for disabling BCG does not match the code.

## Before merge

**1. The tier decision can diverge across TP ranks.** [`_admit_active_schedule` swallows the allocation failure and `return 0`](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py#L1449) per rank, so once one rank's pinned allocation fails, its `host_cache` contents and `current_bytes` no longer match the others. The next time that schedule arrives, the ranks that hit issue no collectives while the ranks that miss issue [`num_layers + 1` all-gathers](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py#L1309) into the same TP group. Faking TP=2 and injecting one allocation failure reproduces it (2-layer arch here; 50 layers in production means 51):

```
admit A: host entries=1   admit B: host entries=0
replay A -> rank0 tier='host_hit'  all_gathers=0
replay A -> rank1 tier='rebuild'   all_gathers=3
```

The tier decision has to be identical on every rank, so it would be better if admission had no per-rank failure branch at all. `is_pin_memory_available()` returning different values per rank triggers the same thing.

<details>
<summary>Reproducer — single process, no multi-GPU needed</summary>

Drop in `python/sglang/multimodal_gen/test/unit/test_minimax_h3_adaln_tp_divergence.py` and run `pytest` on it.

```python
# SPDX-License-Identifier: Apache-2.0

"""Repro for the per-rank host-admission divergence in the tiered AdaLN cache.

A host hit issues no collectives; a rebuild issues ``num_layers + 1``
all-gathers. The tier decision therefore has to be identical on every TP rank.
``_admit_active_schedule`` swallows a per-rank allocation failure, so one rank
can hold a schedule the others do not, and the next request that reuses that
schedule makes the ranks issue different collectives.
"""

from pathlib import Path

import torch
from safetensors.torch import save_file

from sglang.multimodal_gen.configs.models.dits.minimax_h3 import (
    MINIMAX_H3_ADALN_MODALITY_NUM,
    MiniMaxH3DiTArchConfig,
)
from sglang.multimodal_gen.runtime.distributed.parallel_state import (
    maybe_init_distributed_environment_and_model_parallel,
    model_parallel_is_initialized,
)
from sglang.multimodal_gen.runtime.models.dits import minimax_h3 as h3
from sglang.multimodal_gen.test.single_test_file.component_accuracy.utils import (
    ensure_distributed_env_defaults,
)

_TP_SIZE = 2
_ARCH = MiniMaxH3DiTArchConfig(num_layers=2, hidden_size=4, time_embed_dim=3)
_BLOCK_WIDTH = 6 * MINIMAX_H3_ADALN_MODALITY_NUM * _ARCH.hidden_size
_FINAL_WIDTH = 2 * _ARCH.hidden_size
# One all-gather per block plus one for the final layer.
_REBUILD_COLLECTIVES = _ARCH.num_layers + 1


def _write_weights(path: Path) -> None:
    tensors: dict[str, torch.Tensor] = {}
    for layer in range(_ARCH.num_layers):
        prefix = f"blocks.{layer}.adaln_proj.linear"
        tensors[f"{prefix}.weight"] = torch.zeros(_BLOCK_WIDTH, _ARCH.time_embed_dim)
        tensors[f"{prefix}.bias"] = torch.zeros(_BLOCK_WIDTH)
    prefix = "final_layer.adaln_proj.linear"
    tensors[f"{prefix}.weight"] = torch.zeros(_FINAL_WIDTH, _ARCH.time_embed_dim)
    tensors[f"{prefix}.bias"] = torch.zeros(_FINAL_WIDTH)
    save_file(tensors, path)


def _embed(timesteps: torch.Tensor) -> torch.Tensor:
    return timesteps[:, None].expand(-1, _ARCH.time_embed_dim)


def _rank_cache(weight_path: Path) -> h3.MiniMaxH3AdalnCache:
    if not model_parallel_is_initialized():
        ensure_distributed_env_defaults()
        maybe_init_distributed_environment_and_model_parallel(tp_size=1, sp_size=1)
    cache = h3.MiniMaxH3AdalnCache(
        _ARCH,
        weight_files=[str(weight_path)],
        max_plans=4,
        max_plan_width=1,
        host_cache_max_bytes=1 << 20,
    )
    cache.load(torch.device("cpu"))
    return cache


class _CollectiveCounter:
    """Stands in for the TP all-gather so the rebuild path can be counted."""

    def __init__(self) -> None:
        self.calls = 0

    def __call__(self, tensor: torch.Tensor, dim: int = -1) -> torch.Tensor:
        self.calls += 1
        return torch.cat([tensor] * _TP_SIZE, dim=dim)


def _fake_tp(monkeypatch, counter: _CollectiveCounter) -> None:
    monkeypatch.setattr(h3, "get_tp_world_size", lambda: _TP_SIZE)
    monkeypatch.setattr(h3, "get_tp_rank", lambda: 0)
    monkeypatch.setattr(h3, "tensor_model_parallel_all_gather", counter)


def _deny_host_allocation(monkeypatch) -> None:
    """Make the pinned-host allocation inside _admit_active_schedule fail."""
    real_empty = torch.empty

    def fake_empty(*args, **kwargs):
        if kwargs.get("device") == "cpu" and kwargs.get("dtype") is torch.bfloat16:
            raise RuntimeError("simulated host allocation failure")
        return real_empty(*args, **kwargs)

    monkeypatch.setattr(torch, "empty", fake_empty)


def test_host_admission_failure_diverges_tier_across_tp_ranks(tmp_path, monkeypatch):
    weight_path = tmp_path / "model.safetensors"
    _write_weights(weight_path)
    plan_a = torch.tensor([1.0])
    plan_b = torch.tensor([2.0])

    counter_a = _CollectiveCounter()
    counter_b = _CollectiveCounter()
    rank_a = _rank_cache(weight_path)
    rank_b = _rank_cache(weight_path)

    with monkeypatch.context() as m:
        _fake_tp(m, counter_a)
        rank_a.build([plan_a], embed=_embed)
    with monkeypatch.context() as m:
        _fake_tp(m, counter_b)
        _deny_host_allocation(m)
        rank_b.build([plan_a], embed=_embed)

    assert len(rank_a.host_cache) == 1
    assert len(rank_b.host_cache) == 0, "the two ranks already disagree"

    for cache, counter in ((rank_a, counter_a), (rank_b, counter_b)):
        with monkeypatch.context() as m:
            _fake_tp(m, counter)
            cache.build([plan_b], embed=_embed)

    before_a, before_b = counter_a.calls, counter_b.calls
    with monkeypatch.context() as m:
        _fake_tp(m, counter_a)
        access_a = rank_a.build([plan_a], embed=_embed)
    with monkeypatch.context() as m:
        _fake_tp(m, counter_b)
        access_b = rank_b.build([plan_a], embed=_embed)

    assert access_a.tier == "host_hit"
    assert access_b.tier == "rebuild"
    # The mismatch NCCL would see: one rank posts nothing, the other posts
    # num_layers + 1 all-gathers into the same TP group.
    assert counter_a.calls - before_a == 0
    assert counter_b.calls - before_b == _REBUILD_COLLECTIVES
```

</details>

**2. The default configuration is a regression against today.** [`host_cache_gb` defaults to `0.0`](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/server_args/server_args.py#L314), which disables the host tier, while [`max_plans` still defaults to `64`](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/server_args/server_args.py#L309) and `plan_width` to 4, so the workspace still costs 2.31 GiB. The GPU side meanwhile goes from "accumulate up to 64 plans across requests" to "invalidate everything on a new schedule". At defaults that is no memory saved and no cross-request hits: the A->B->A third request in #35495 goes from a 0.2 ms GPU hit to a rebuild. The 7.392 ms row in the PR description was measured with `host-cache-gb > 0`, `max-plans 7` and `plan-width 3` — could you state that configuration in the benchmark table?

**3. The stated reason for disabling BCG does not match the code.** [`bool(matches.any())` inside `lookup()`](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py#L1470) is a device-to-host sync that capture rejects, and [online and sidecar take the same `lookup()` in the forward](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py#L2492). By that reasoning the sidecar path should be disabled too; if the actual cause is something else, it would help to spell it out at the [downgrade](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/server_args/server_args.py#L675). Separately, that check uses `type(getattr(...)).__name__`, while `.claude/rules/no-getattr-defensive.md` asks for an `isinstance` check against the imported class.

## Suggestions

4. [`schedule_key = tuple(wanted)`](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py#L1242) takes dict insertion order, so the same plan set discovered in a different order is a different key and costs another rebuild. `unique_timesteps` already comes out of `torch.unique(sorted=True)`, so an individual plan key is canonical; only the schedule level needs normalizing.
5. [The shard GEMM all-gathers immediately](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py#L1309), so what lands in host memory is full width — at TP=8 that is 8 ranks holding identical data, replicated again across SP ranks. Storing the TP-local shard and all-gathering after H2D divides both host bytes and H2D bytes by tp.
6. The two new `@dataclass` definitions: `.claude/rules/no-dataclasses.md` asks for `msgspec.Struct` in new data containers.
7. There is no TP>1 test case, and item 1 only shows up at TP>1.

## For reference, not for this PR

- [`_invalidate_workspace`](https://github.com/sgl-project/sglang/blob/49c012d33e6f2799ba245b07f8410ccc6117f130/python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py#L1364) clears `plan_lengths` but not `block_params`, so for a schedule mixing plan widths the trailing columns of the shorter plans carry the previous schedule's bytes into the host entry. `block()` only reads `:num_timesteps` today, so nothing reads them and this is not a bug — but one change to that slicing turns it into a silent numerical error.
- `h2d_transfers=1` undercounts: `_load_host_entry` issues three `copy_` calls, and when `width < max_plan_width` the destination slices are non-contiguous and stage additionally.
- `MiniMaxH3AdalnAccess` puts per-request deltas (`h2d_bytes`) and process-cumulative counters (`rebuilds`) in one flat namespace, so the perf dump mixes the two.
